types/scheduler/tracing.d.ts

```diff
@@ -0,0 +1,131 @@
+ * The default behavior for the types is to use a union of all possible types.
+ */
+// tslint:disable-next-line: no-empty-interface
+export interface Build {}
+
+export type EnableSchedulerTracing = Build extends { type: infer BuildType }
+  ? BuildType extends "production" | "profiling"
+    ? false
+    : BuildType extends "development"
+    ? true
+    : undefined
+  : undefined;
+
+type TypeByBuildFlag<
+  Flag extends boolean | undefined,
+  WhenTrue,
+  WhenFalse
+> = Flag extends undefined
+  ? (WhenTrue | WhenFalse)
+  : Flag extends true
+  ? WhenTrue
+  : WhenFalse;
+
+type IfSchedulerTracing<WhenTrue, WhenFalse> = TypeByBuildFlag<
+  EnableSchedulerTracing,
+  WhenTrue,
+  WhenFalse
+>;
+
+export interface Interaction {
+  __count: number;
+  id: number;
+  name: string;
+  timestamp: number;
+}
+
+export interface Subscriber {
+  /**
+   * A new interaction has been created via the trace() method.
+   */
+  onInteractionTraced: (interaction: Interaction) => void;
+
+  /**
+   * All scheduled async work for an interaction has finished.
+   */
+  onInteractionScheduledWorkCompleted: (interaction: Interaction) => void;
+
+  /**
+   * New async work has been scheduled for a set of interactions.
+   * When this work is later run, onWorkStarted/onWorkStopped will be called.
+   * A batch of async/yieldy work may be scheduled multiple times before completing.
+   * In that case, onWorkScheduled may be called more than once before onWorkStopped.
+   * Work is scheduled by a "thread" which is identified by a unique ID.
+   */
+  onWorkScheduled: (interactions: Set<Interaction>, threadID: number) => void;
+
+  /**
+   * A batch of scheduled work has been canceled.
+   * Work is done by a "thread" which is identified by a unique ID.
+   */
+  onWorkCanceled: (interactions: Set<Interaction>, threadID: number) => void;
+
+  /**
+   * A batch of work has started for a set of interactions.
+   * When this work is complete, onWorkStopped will be called.
+   * Work is not always completed synchronously; yielding may occur in between.
+   * A batch of async/yieldy work may also be re-started before completing.
+   * In that case, onWorkStarted may be called more than once before onWorkStopped.
+   * Work is done by a "thread" which is identified by a unique ID.
+   */
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void;
+
+  /**
+   * A batch of work has completed for a set of interactions.
+   * Work is done by a "thread" which is identified by a unique ID.
+   */
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void;
+}
+
+export interface InteractionsRef {
+  current: Set<Interaction>;
+}
+
+export interface SubscriberRef {
+  current: Subscriber | null;
+}
+
+export const __interactionsRef: IfSchedulerTracing<InteractionsRef, null>;
+export const __subscriberRef: IfSchedulerTracing<SubscriberRef, null>;
+
+export function unstable_clear<T>(callback: () => T): T;
```

Comment:
Sorry, one last question. Since this function has a try/catch block around the callback and doesn't return anything if it catches an error, should the return type be `T | undefined`? This is also true of a couple other functions.

Reply:
It has a try-finally. The error isn't caught. I had to get used to this as well. I had a bunch of `| undefined` types in place until I realized that they are never caught. `finally` is just a cleanup construct.